indigo_app/js/components/DocumentTOCView.vue

```diff
@@ -0,0 +1,297 @@
+
+      this.toc = toc;
+      this.roots = roots;
+
+      this.mergeIssues();
+    },
+
+    issuesChanged () {
+      this.mergeIssues();
+      this.render();
+    },
+
+    mergeIssues () {
+      // fold document issues into the TOC
+      const withIssues = [];
+
+      _.each(this.toc, (entry) => {
+        entry.issues = [];
+      });
+
+      this.issues.each((issue) => {
+        // find the toc entry for this issue
+        const element = issue.get('element');
+
+        if (element) {
+          const entry = this.entryForElement(element);
+          if (entry) {
+            entry.issues.push(issue);
+            withIssues.push(entry);
+          }
+        }
+      });
+
+      // now attach decent issue descriptions
+      _.each(withIssues, (entry) => {
+        let severity = _.map(entry.issues, (issue) => { return issue.get('severity'); });
+        severity = _.contains(severity, 'error') ? 'error' : (_.contains(severity, 'warning') ? 'warning': 'information');
+
+        entry.issues_title = entry.issues.length + ' issue' + (entry.issues.length === 1 ? '' : 's');
+        entry.issues_description = entry.issues.map((issue) => { return issue.get('message'); }).join('<br>');
+        entry.issues_severity = severity;
+      });
+    },
+
+    entryForElement (element) {
+      // find the TOC entry for an XML element
+      const tradition = Indigo.traditions.get(this.backBoneContext.model.document.get('country')),
+          toc = this.toc;
+
+      // first, find the closest element's ancestor that is a toc element
+      while (element) {
+        if (tradition.is_toc_element(element)) {
+          // now get the toc item for this element
+          for (let i = 0; i < toc.length; i++) {
+            if (toc[i].element === element) return toc[i];
+          }
+        }
+
+        element = element.parentElement;
+      }
+    },
+
+    render () {
+      const formatItems = (item) => {
+        const rightIcon = (() => {
+          if(item.issues && item.issues.length) {
+            const icon = document.createElement('i');
+            icon.className = 'float-right issue-icon issue-' + item.issues_severity;
+            icon.dataset.toggle = 'popover';
+            icon.dataset.content = item.issues_description;
+            icon.dataset.title = item.issues_title;
+            icon.dataset.trigger = 'hover';
+            icon.dataset.placement = 'bottom';
+            icon.dataset.html = true;
+            return icon.outerHTML;
+          }
+          return null;
+        })();
+        return ({
+          title: item.title,
+          children: item.children && item.children.length ? item.children.map(formatItems) : [],
+          rightIcon,
+          index: item.index,
+          selected: item.selected,
+        });
+      }
+
+      this.tocItems = this.roots.map(formatItems);
+      $('#toc [data-toggle="popover"]').popover();
+      this.$refs.tocController.expandAll();
```

Comment:
This can move into `mounted` because it should only need to be done right at the start.

Reply:
done